examples/bert/bert_classifier_main_V2.py

```diff
@@ -0,0 +1,257 @@
+
+    vars_with_decay = []
+    vars_without_decay = []
+    for name, param in model.named_parameters():
+        if 'layer_norm' in name or name.endswith('bias'):
+            vars_without_decay.append(param)
+        else:
+            vars_with_decay.append(param)
+
+    opt_params = [
+        {
+            'params': vars_with_decay,
+            'weight_decay': 0.01,
+        },
+        {
+            'params': vars_without_decay,
+            'weight_decay': 0.0,
+        }
+    ]
+    optim = BertAdam(opt_params, betas=(0.9, 0.999), eps=1e-6, lr=static_lr)
+
+    scheduler = torch.optim.lr_scheduler.LambdaLR(
+        optim, functools.partial(
+            model_utils.get_lr_multiplier, total_steps=num_train_steps,
+            warmup_steps=num_warmup_steps))
+
+    train_dataset = tx.data.RecordData(hparams=config_data.train_hparam)
+    eval_dataset = tx.data.RecordData(hparams=config_data.eval_hparam)
+    test_dataset = tx.data.RecordData(hparams=config_data.test_hparam)
+
+    iterator = tx.data.DataIterator(
+        {"train": train_dataset, "eval": eval_dataset, "test": test_dataset}
+    )
+
+    def _train_epoch():
+        """Trains on the training set, and evaluates on the dev set
+        periodically.
+        """
+        iterator.switch_to_dataset("train")
+
+        model.train()
+        for batch in iterator:
+            optim.zero_grad()
+            input_ids = batch["input_ids"]
+            segment_ids = batch["segment_ids"]
+            labels = batch["label_ids"]
+
+            if torch.cuda.is_available():
+                input_ids = input_ids.cuda()
+                segment_ids = segment_ids.cuda()
+                labels = labels.cuda()
+
+            input_length = (1 - (input_ids == 0).int()).sum(dim=1)
+
+            logits, _ = model(input_ids, input_length, segment_ids)
+
+            if model.is_binary:
+                loss = F.binary_cross_entropy(logits.view(-1),
+                                              labels.view(-1),
+                                              reduction='mean')
+            else:
+                loss = F.cross_entropy(logits.view(-1, model.num_classes),
+                                       labels.view(-1),
+                                       reduction='mean')
+
+            loss.backward()
+            optim.step()
+            scheduler.step()
+            step = scheduler.last_epoch
+
+            dis_steps = config_data.display_steps
+            if dis_steps > 0 and step % dis_steps == 0:
+                logging.info("step: %d; loss: %f", step, loss)
+
+            eval_steps = config_data.eval_steps
+            if eval_steps > 0 and step % eval_steps == 0:
+                _eval_epoch()
+
+    @torch.no_grad()
+    def _eval_epoch():
+        """Evaluates on the dev set.
+        """
+        iterator.switch_to_dataset("eval")
+        model.eval()
+        cum_acc = 0.0
+        cum_loss = 0.0
+        nsamples = 0
```

Comment:
We can use `texar.utils.AverageRecorder` instead of manually keeping track of these.